### DISCUSSION TITLE
🖐️ Running a `go test` container and calling test containers fail to reach nested containers

### DISCUSSION THREAD

**Discussion by vic3lord** [dagger/dagger#12094] [Help]
### What happened? What did you expect to happen?

I'm sharing the docker.sock with the `go test` container.

Locally on docker desktop it works if I add `TC_HOST=host.docker.internal` but on github actions it does not work.

**Comment by grouville** [comment]
Hi @vic3lord,
I'm not sure to fully understand 👼 . Can you please detail the issue / paste a repro here so that we can isolate the root cause as fast as possible 😇 

Also, could you also reference your github action workflow ?

**Comment by vic3lord** [comment]
Hey @grouville,
They issue is that host.docker.internal isn't available for some reason on Github actions but it does exist on "docker for desktop"
- My github action is just calling the go binary dagger workflow.
- It's trying to access the postgres container but it's timing out due to the issue with host networking.
- I setup the postgres container via `_test.go` file and https://golang.testcontainers.org

**Comment by vic3lord** [comment]
I moved away from test containers to dagger service containers just for CI (although it's a bit of a pain to manage different versions of code between `go test` to when it's a dagger workflow)

**Comment by grouville** [comment]
> dagger

Nice, I was to propose that. The other things that should work (I was about to explore tomorrow, as it's easter today), is to rely on Colima as your container runtime.
Following their docs, it totally seems compatible: https://github.com/abiosoft/colima/blob/65ee3d284b81bfebc09783f5e35a186584aa4bc1/docs/FAQ.md?plain=1#L148-L161

On CI: https://github.com/dagger/dagger/blob/4be9d6ee11cb5d3a66ec3e0f846436535009fd5e/.github/workflows/publish.yml#L110-L116

It would require some testing (did not test it thoroughly) but it exposes a socket, and it would be consitent on CI and locally, at the cost of performance, as it relies on QEMU: 
<img width="466" alt="image" src="https://user-images.githubusercontent.com/31691250/230887663-83eb11c0-8ab6-44fd-ba31-c798c2b48da2.png">

**Comment by jpadams** [comment]
cc @vito @d3rp3tt3

**Comment by mircubed** [comment]
reopening to ensure that we solved for @vic3lord 's problem. 

@vic3lord is there anything else that you'd like to see from Dagger to make this experience easier? Is there documentation that you'd expect to see for this workaround? 

> I moved away from test containers to dagger service containers just for CI (although it's a bit of a pain to manage different versions of code between go test to when it's a dagger workflow)

@vikram-dagger , I am adding a docs tag since I think we should document the workaround somewhere in our docs.

**Comment by vic3lord** [comment]
@mircubed I'm not sure changing the runtime to  colima is a good alternative for the basic docker we already get in github actions.

- We either need a way to get the docker gateway using the SDK (which isn't available in docker for linux it's a trick for docker for mac/windows only)
- Or we should be able to run long running containers for development and `dagger services` alongside those for TDD cycles without killing those at the end

**Comment by gerhard** [comment]
If I understand correctly, you want to run integration tests with https://golang.testcontainers.org/ in GitHub Actions using Dagger. This requires for the Docker instance running in GitHub Actions to be available to the container running within Dagger. This diagram captures what I described:

```mermaid
graph TD
    subgraph "GitHub Actions"
        subgraph "Docker"
            socket(unix:///var/run/docker.sock)
            subgraph "Dagger"
                go-test
            end
            postgres
        end
    end

    socket --> |mount| go-test
    go-test --> |start / stop| postgres((postgres container))
```

While that would satisfy the initial requirement, I am wondering where does Dagger's Service Containers feature fall short @vic3lord? In other words, why does this diagram not work for you?

```mermaid
graph TD
    subgraph "GitHub Actions"
        subgraph "Docker"
            subgraph "Dagger"
                go-test
                postgres
            end
        end
    end
    go-test --> postgres((postgres container))
```

FTR:
- https://docs.dagger.io/757394/use-service-containers/
- https://dagger.io/blog/dagger-engine-0-4-0

cc @vito

**Comment by gerhard** [comment]
When we solve this one, it would be worth capturing it as a recipe in the Cookbook @vikram-dagger.

**Comment by paul-prdca** [comment]
I'm running into this same issue, let me see if I can clarify my use case.

I've got code running with a "testcontainers" container (in my case Vault).  While I could probably adjust this to use Dagger service containers, it works currently with `go test` outside of Dagger.   Most importantly to me, that means it works easily through my IDE (where I'd like to get quick response for testing and coverage).

Rearranging the IDE test process to require Dagger and plumbing the coverage results back out (especially in a case where I'm testing a single function) seems like work for little reward.  That's _before_ the issue reported here where it won't work in Github Actions, my eventual CI target.

So, for flow reasons I'd _like_ it to continue to work with `go test` but be able to wrap that easily in Dagger when I have a full pipeline to test/build/publish.  Right now I'm looking at doing testing outside Dagger, and multi-arch build/publish with Dagger after a testing phase completes.

**Comment by king-11** [comment]
I still think this is still an issue am not able to run testcontainer test inside of dagger runtime, which spawn up `postgres` and `etcd` containers.

I am using `colima` and `macOS` M4. The error I see is related to cgroups, I tried the approach mentioned here https://github.com/dagger/dagger/issues/5292#issuecomment-1593750070 but for some reason the entrypoint doesn't really get picked up, not sure what am I missing over here

```python
ENTRYPOINT_CGROUP_V2 = """#!/bin/sh

set -o errexit
set -o nounset

#########################################################################################################################################
# DISCLAIMER																																																														#
# Copied from https://github.com/moby/moby/blob/ed89041433a031cafc0a0f19cfe573c31688d377/hack/dind#L28-L37															#
# Permission granted by Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp> (https://github.com/k3d-io/k3d/issues/493#issuecomment-827405962)	#
# Moby License Apache 2.0: https://github.com/moby/moby/blob/ed89041433a031cafc0a0f19cfe573c31688d377/LICENSE														#
#########################################################################################################################################
if [ -f /sys/fs/cgroup/cgroup.controllers ]; then
  echo "[$(date -Iseconds)] [CgroupV2 Fix] Evacuating Root Cgroup ..."
	# move the processes from the root group to the /init group,
  # otherwise writing subtree_control fails with EBUSY.
  mkdir -p /sys/fs/cgroup/init
  busybox xargs -rn1 < /sys/fs/cgroup/cgroup.procs > /sys/fs/cgroup/init/cgroup.procs || :
  # enable controllers
  sed -e 's/ / +/g' -e 's/^/+/' <"/sys/fs/cgroup/cgroup.controllers" >"/sys/fs/cgroup/cgroup.subtree_control"
  echo "[$(date -Iseconds)] [CgroupV2 Fix] Done"
fi

exec "$@"
"""


    def dind_service(self) -> Service:
      return (dag
        .container()
        .from_("docker:dind")
        .with_new_file("/usr/bin/entrypoint.sh", ENTRYPOINT_CGROUP_V2, permissions = 755)
        .with_env_variable("TINI_SUBREAPER", "")
        .with_env_variable("DOCKER_TLS_CERTDIR", "")
        .with_exposed_port(2375)
        .with_entrypoint(["/usr/bin/entrypoint.sh"])
        .as_service(args=["dockerd", "--tls=false", "--host=tcp://0.0.0.0:2375"], insecure_root_capabilities=True)
      )

    def with_dind_service(self, container: Container) -> Container:
      return (container
                .with_service_binding("docker", self.dind_service())
                .with_env_variable("DOCKER_HOST", "tcp://docker:2375")
                .with_env_variable("TESTCONTAINERS_RYUK_DISABLED", "true")
              )

```

The error I am seeing is as below
```
time="2025-04-21T14:13:05.247275588Z" level=info msg="API listen on [::]:2375"
time="2025-04-21T14:14:16.694993301Z" level=error msg="failed to enable controllers ([cpuset cpu io memory hugetlb pids rdma misc])" error="failed to write subtree controllers [cpuset cpu io memory hugetlb pids rdma misc] to \"/sys/fs/cgroup/docker/cgroup.subtree_control\": write /sys/fs/cgroup/docker/cgroup.subtree_control: no such file or directory"
time="2025-04-21T14:14:16.695775600Z" level=warning msg="error from *cgroupsv2.Manager.EventChan" error="failed to add inotify watch for \"/sys/fs/cgroup/docker/96ae191bbe8a76e431bae3a491a6cd76e99f4e867e07486e1bbd13d5c1626898/memory.events\": no such file or directory"
time="2025-04-21T14:14:18.820503598Z" level=error msg="failed to enable controllers ([cpuset cpu io memory hugetlb pids rdma misc])" error="failed to write subtree controllers [cpuset cpu io memory hugetlb pids rdma misc] to \"/sys/fs/cgroup/docker/cgroup.subtree_control\": write /sys/fs/cgroup/docker/cgroup.subtree_control: no such file or directory"
time="2025-04-21T14:14:18.820866018Z" level=warning msg="error from *cgroupsv2.Manager.EventChan" error="failed to add inotify watch for \"/sys/fs/cgroup/docker/81f38080498e28a6faaaa3b1ebd772ccd4d8333a7936d49c0764cdae10272a23/memory.events\": no such file or directory"
```

The container processes aren't able to connect with the `testcontainers` as they never get spun up
